### DISCUSSION TITLE
How to ingest data from ForgeRock app --OpenTelemetry(Java) -- Splunk

### DISCUSSION THREAD

**Discussion by Rockstar2107** [open-telemetry/opentelemetry-java#3794] [Q&A]
How to ingest data from ForgeRock API --> OpenTelemetry(Java) --> Splunk? Can anyone throw me some idea or steps to kick start the integration.

**Comment by anuraaga** [comment]
Hi @Rockstar2107 - could you describe how you access the ForgeRock API? I tried looking around a bit but couldn't find the Java SDK for it. Normally we'd expect registering an interceptor to it if it supports it, or otherwise wrapping it, to create spans around the API calls. Then you could point the OTLP exporter to the Splunk endpoint to ingest them.

**Comment by Rockstar2107** [comment]
Hi Anuraaga, thanks for responding, hey I had gone through this link in ForgeRock document on how to record the audit data into splunk.

https://backstage.forgerock.com/docs/ig/7/maintenance-guide/audit-splunk.html

Where they have used the Splunk HEC token and provided token details in the 30-splunk.json. 

"Normally we'd expect registering an interceptor to it if it supports it, or otherwise wrapping it, to create spans around the API calls" -- I did not understand this statement so could you please explain me what you mean by "interceptor" and OTLP exporter.

**Reply by anuraaga** [reply]
Thanks for the link - is Identity Gateway a binary you run? In that case I guess you don't have any actual code to add OpenTelemetry to - otherwise, if you were to actually build an Identity Gateway with custom code, then an OpenTelemetry AuditEventHandler could probably be implemented for generating telemetry.

https://backstage.forgerock.com/docs/ig/7/apidocs/org/forgerock/audit/events/handlers/AuditEventHandler.html

It seems that guide describes how to ingest IG events into splunk directly and is probably the way to do it and wouldn't need adding OpenTelemetry - is there something else you're trying to achieve by adding OpenTelemetry to the mix?

**Comment by Rockstar2107** [comment]
Hey actually we wanted to get the ForgeRock (Metrics, Traces and Event logs) in to splunk but there is  additional application called "open telemetry" between ForgeRock and Splunk, client wanted to see whatever the details captured from ForgeRock to open telemetry should see the same details in Splunk.

**Comment by Rockstar2107** [comment]
We have already implemented the same into New Relic application using the Jar file and it is working fine. 
https://libraries.io/maven/com.newrelic.telemetry:newrelic-opentelemetry-javaagent

**Reply by jkwatson** [reply]
Have you tried using Splunk's OpenTelemetry distribution at https://github.com/signalfx/splunk-otel-java?

**Reply by Rockstar2107** [reply]
Hi Watson,

thanks for your response on this, actually client wanted to implement this in the splunk enterprise but when we go splunk-otel-java connector then we need to go for the Splunk observability cloud add-on which is additional cost to the company. So could please tell me whether we can use this connecter with the splunk enterprise version.

**Reply by jkwatson** [reply]
So, are you asking for an add-on for the splunk-otel-connector that will send data to Splunk Enterprise? If so, this is definitely not the right place to ask for that. I would open an issue/discussion at https://github.com/signalfx/splunk-otel-collector for this support.

**Reply by Rockstar2107** [reply]
No I am not asking any add-on for Splunk-otel-connector that will send data to Splunk Enterprise?  whether we need to buy the Splunk observability cloud module in-order to ingest the open telemetry data as in the link 
https://github.com/signalfx/splunk-otel-collector -- under  ​README.md it is mentioned like this that Splunk OpenTelemetry Connector "Splunk-otel-connector"  is a distribution of the OpenTelemetry Collector. It provides a unified way to receive, process, and export metric, trace, and log data for Splunk Observability Cloud. 

I want to know how to ingest data from ForgeRock application data  ---> Open telemetry app --> To Splunk Enterprise ? Kindly throw me some idea on the same.

**Reply by jkwatson** [reply]
Without the collector in between your otel app and SE, I don't think there's currently a way to do it, unless you want to write a custom HEC exporter for Java. I don't think this is something on our (Splunk's) roadmap at the moment.

**Reply by Rockstar2107** [reply]
Hi watson,

Good Morning, Could you please share us the steps on how to ingest open telemetry data into splunk.
Currently we have integrated ForgeRock and open telemetry in our test lab machine and able to see the data in the tomact console.

Integration steps (ForgeRock/Open telemetry) steps

1) Downloaded ForgeRock version 7 and apache  tomcat 9.044 and installed in windows machine.
2) Downloaded Open Telemetry jar file and installed in the same windows machine.
3) In Tomcat console we are able to see the open Telemetry data.

Now we wanted to ingest the same data from open telemetry into splunk enterprise, so could please throw some lights on how to ingest the data into splunk.

**Reply by jkwatson** [reply]
I don't personally know how to do this, but I will ask for assistance internally.

**Reply by jkwatson** [reply]
@Rockstar2107 What company do you work for, so we can find the appropriate resource to help you out with this? Thanks!

**Reply by Rockstar2107** [reply]
I am working in Infosys and this POC is our of financial client.

**Reply by jkwatson** [reply]
Great. I'm finding you a resource at Splunk who can assist with your particular case.

**Reply by Rockstar2107** [reply]
Hi Watson,

Good Morning, hey what are the pre-requisites and pre task which needs to considered for getting the open telemetry data into Splunk.

**Reply by jkwatson** [reply]
@Rockstar2107 I'm definitely not an expert here (I'm still tracking down a Splunk internal resource for you). But, I think the basic steps are these:
1) Install the otel collector with the splunk HEC exporter configured to export to your Splunk Enterprise instance
2) Point your java agent/sdk at the collector for all OTLP exports
3) that's it!
